jmx-metrics/src/main/groovy/io/opentelemetry/contrib/jmxmetrics/InstrumentHelper.groovy

```diff
@@ -52,140 +56,201 @@ class InstrumentHelper {
-        // update operations in a list of closures per instrument to be executed after all values
-        // are established, potentially as a single updater.  This is done because a single MBeanHelper
-        // can represent multiple MBeans (each with different values for an attribute) and the labelFuncs
-        // will create multiple datapoints from the same instrument identifiers.
-        def tupleToUpdates = [:] // tuple is of form (instrument, instrumentName, description, unit)
-
-        values.each { collectedValue ->
-            def mbean = collectedValue.getV1()
-            def attribute = collectedValue.getV2()
-            def value = collectedValue.getV3()
-            if (value instanceof CompositeData) {
-                value.getCompositeType().keySet().each { key ->
-                    def val = value.get(key)
-                    def updatedInstrumentName = "${instrumentName}.${key}"
-                    def labels = getLabels(mbean, labelFuncs, mBeanAttributes[attribute])
-                    def tuple = new Tuple(instrument, updatedInstrumentName, description, unit)
-                    logger.fine("Recording ${updatedInstrumentName} - ${instrument.method} w/ ${val} - ${labels}")
-                    if (!tupleToUpdates.containsKey(tuple)) {
-                        tupleToUpdates[tuple] = []
-                    }
-                    tupleToUpdates[tuple].add(prepareUpdateClosure(instrument, val, labels))
+        mBeanAttributes.keySet().each { attribute ->
+            try {
+                // Look at the collected mbeans to evaluate if the attributes requested are
+                // composite data types or simple. Composite types require different parsing to
+                // end up with multiple recorders in the same callback.
+                def keySet = getCompositeKeys(attribute, mbeans)
+                if (keySet.size() > 0) {
+                    compositeAttributes.add(new Tuple2<String, Set<String>>(attribute, keySet))
+                } else {
+                    simpleAttributes.add(attribute)
                 }
-            } else if (value != null) {
-                def labels = getLabels(mbean, labelFuncs, mBeanAttributes[attribute])
-                def tuple = new Tuple(instrument, instrumentName, description, unit)
-                logger.fine("Recording ${instrumentName} - ${instrument.method} w/ ${value} - ${labels}")
-                if (!tupleToUpdates.containsKey(tuple)) {
-                    tupleToUpdates[tuple] = []
-                }
-                tupleToUpdates[tuple].add(prepareUpdateClosure(instrument, value, labels))
+            } catch (AttributeNotFoundException ignored) {
+                logger.fine("Attribute ${attribute} not found on any of the collected mbeans")
+            } catch (InvalidAttributeValueException ignored) {
+                logger.info("Attribute ${attribute} was not consistently CompositeData for " +
```

Comment:
I think this may warrant a ~breaking change changelog entry or readme update in the chance that users run into it

Reply:
I added a note to the readme, from what I can see it looks like changelog entries are added outside the scope of the PR introducing the change. If I'm wrong though, I'm happy to add whatever's required for a changelog entry.